silq/tools/circuit_tools.py

```diff
@@ -0,0 +1,175 @@
+                raise RuntimeError(f'Cannot understand gate {gate}')
+
+    elif isinstance(circuit, str):
+        # Replace [] by identity gate
+        expanded_circuit = circuit.replace('[]', 'Gi')
+
+        # Expand any expressions ({gates})^exponent
+        expand_subcircuit = lambda match: match.group(0).replace(
+            match.group(0), match.group(1) * int(match.group(2))
+        )
+        expanded_circuit = re.sub(r'\(([A-Za-z]+)\)\^([0-9]+)', expand_subcircuit, expanded_circuit)
+
+        # Expand any expressions {gate}^exponent (notice missing parentheses
+        expand_subcircuit = lambda match: match.group(0).replace(
+            match.group(0), ('G'+match.group(1)) * int(match.group(2))
+        )
+        expanded_circuit = re.sub(r'G([a-z0-9:]+)\^([0-9]+)', expand_subcircuit, expanded_circuit)
+
+        # Remove all single parentheses without exponent
+        expanded_circuit = re.sub(r"\(|\)", "", expanded_circuit)
+
+        # Newer circuits end with @(qubit idxs), which should be removed
+        expanded_circuit = expanded_circuit.split('@')[0]
+
+        if expanded_circuit == '{}':
+            expanded_circuit = ''
+
+        # Split by every letter G
+        expanded_circuit = ['G'+ gate for gate in expanded_circuit.split('G')[1:]]
+    elif isinstance(circuit, list):
+        expanded_circuit = circuit
+    else:
+        raise RuntimeError("circuit type not understood")
+
+    if target_type == str:
+        if expanded_circuit:
+            return ''.join(expanded_circuit)
+        else:
+            return '{}'
+    elif target_type == list:
+        return expanded_circuit
+    else:
+        raise NameError(f'Cannot select target type {target_type}')
+
+
+def save_circuits(circuits, filepath):
+    """Save list of circuits to a .txt file"""
+    filepath = Path(filepath)
+    if not filepath.suffix:
+        filepath = filepath.with_suffix('.txt')
+
+    with filepath.open('w') as f:
+        for circuit in circuits:
+            circuit_str = convert_circuit(circuit, target_type=str)
+            f.write(circuit_str + '\n')
+
+
+def load_circuits(
+        filepath: Union[str, Path],
+        target_type: type = list,
+        load_probabilities: bool = False
+):
+    filepath = Path(filepath)
+    with open(filepath.with_suffix('.txt')) as file:
+        lines = [line.strip() for line in file]
+
+    # Check if line already contains data
+    if lines[0].startswith('##'):
+        # Remove two headers
+        lines = lines[1:]
+
+    # Remove any existing results
+    circuit_strings = [line.split(' ')[0] for line in lines]
+
+    # Ensure all strings are of the target type
+    circuits = [
+        convert_circuit(circuit, target_type=target_type)
+        for circuit in circuit_strings
+    ]
+
+    if load_probabilities:
+        # Ignore when there is more than one space
+        lines = [line.replace('   ', ' ') for line in lines]
+        lines = [line.replace('  ', ' ') for line in lines]
```

Comment:
Alternatively, using regex
```suggestion
        lines = [re.sub(' +', ' ', line) for line in lines]
```

Reply:
Hmm, this is ever so slightly slower than the string replace method.
Arguably, it reads better than the 2 lines. Up to you guys in the end.

Final alternative which I just remembered works: `lines = [' '.join(line.split()) for line in lines]`

Reply:
Nice, I like it, fixed